packages/vow/src/rejection-tracker.js

```diff
@@ -0,0 +1,97 @@
+// @ts-check
+import { Fail, X } from '@endo/errors';
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+
+import { sink } from './vow-utils.js';
+
+/**
+ * @import {PromiseKit} from '@endo/promise-kit';
+ * @import {Zone} from '@agoric/base-zone';
+ * @import {VowPayloadV0} from './types.js';
+ */
+
+/**
+ * @param {Promise<void>} [rejected]
+ * @param {PromiseKit<void>} [upgradedPK]
+ */
+const makeUnhandledRejection = (rejected, upgradedPK) => {
+  return () => {
+    // Handle the rejected promise to silence it.
+    rejected?.catch(sink);
+    // Resolve the upgraded promise to prevent it from being rejected by a
+    // future upgrade.
+    upgradedPK?.resolve();
+  };
+};
+
+const VowCapShape = M.remotable('VowCap');
+const VowRejectionTrackerKitI = {
+  public: M.interface('VowRejectionTracker', {
+    handle: M.call(VowCapShape).returns(),
+    reject: M.call(VowCapShape, M.raw(), M.promise()).returns(),
+  }),
+  upgradeRejectionWatcher: M.interface('UpgradeRejectionWatcher', {
+    onRejected: M.call(M.raw(), M.raw()).returns(),
+  }),
+};
+
+/** @param {Zone} zone */
+export const prepareVowRejectionTracker = zone => {
+  /** @type {WeakMap<VowPayloadV0, () => void>} */
+  const vowToCancelUnhandledRejection = new WeakMap();
+
+  const makeVowRejectionTrackerKit = zone.exoClassKit(
+    'VowRejectionTrackerKit',
+    VowRejectionTrackerKitI,
+    () => ({}),
+    {
+      public: {
+        handle(vowV0) {
+          const cancel = vowToCancelUnhandledRejection.get(vowV0);
+          if (!cancel) {
+            throw Fail`vow ${vowV0} already handled`;
+          }
+          cancel();
+        },
+        /**
+         *
+         * @param {VowPayloadV0} vowV0
+         * @param {any} reason
+         * @param {Promise<void>} [rejected]
+         */
+        reject(vowV0, reason, rejected = Promise.reject(reason)) {
```

Comment:
I think this works but I'm somewhat uncomfortable with vows passing its internal promise object.
Consider removing the need to pass the vow's internal promise and create an independent one instead. If we fix the issue that we only call reject if the vow was not "possibly handled" (called shorten on pending state), then vows could create a its internal promise as handled, and rely on the rejection tracker for unhandled reporting.